File: Dijkstra's Algorithm/dijkstra.cpp
#include <iostream>
#define INF 10000
using namespace std;

struct HNode
{
  int key;
  int value;
};

class Heap
{
  HNode* h;
  int size, end;
  int* pos;
public:
  Heap(int n)
  {
    size=n;
    end=n;
    h=new HNode[n];
    pos=new int[n];
    for(int i=0; i<n; i++)
    {
      h[i].key=i;
      h[i].value=INF;
      pos[i]=i;
    }
  }
  void swap(int* a, int* b)
  {
    *a=*a+*b;
    *b=*a-*b;
    *a=*a-*b;
  }
  int empty()
  {
    if(end==0)
    {
      return 1;
    }
    return 0;
  }
  int parent(int i)
  {
    return (((i+1)/2)-1);
  }
  void heapify(int root)
  {
    int min=root;
    if(((2*root)+1)<end && h[((2*root)+1)].value<h[min].value)
    {
      min=(2*root)+1;
    }
    if(((2*root)+2)<end && h[((2*root)+2)].value<h[min].value)
    {
      min=(2*root)+2;
    }
    if(min!=root)
    {
      swap(&h[root].key, &h[min].key);
      swap(&h[root].value, &h[min].value);
      pos[h[root].key]=root;
      pos[h[min].key]=min;
      heapify(min);
    }
  }
  HNode extractmin()
  {
    HNode m=h[0];
    h[0]=h[end-1];
    end=end-1;
    pos[h[0].key]=0;
    heapify(0);
    return m;
  }
  void deckey(int k, int v)
  {
    int i=pos[k];
    if(h[i].value<v)
    {
      return;
    }
    h[i].value=v;
    while(i>0 && h[i].value<h[parent(i)].value)
    {
      swap(&h[i].key, &h[parent(i)].key);
      swap(&h[i].value, &h[parent(i)].value);
      pos[h[i].key]=i;
      pos[h[parent(i)].key]=parent(i);
      i=parent(i);
    }
  }
  void display()
  {
    cout<<"Heap:"<<endl;
    for(int i=0; i<end; i++)
    {
      cout<<h[i].key<<"("<<h[i].value<<")"<<" ";
    }
    cout<<endl<<"Positions:"<<endl;
    for(int i=0; i<size; i++)
    {
      cout<<pos[i]<<" ";
    }
    cout<<endl;
  }
};

struct Node
{
  int data;
  int weight;
  Node* next;
};

Node* Graph;

void vlist(int n)
{
  Graph=new Node[n];
  for(int i=0; i<n; i++)
  {
    Graph[i].data=i;
    Graph[i].weight=0;
    Graph[i].next=NULL;
  }
}

void addedge(int u, int v, int w)
{
  Node* temp=new Node;
  temp->data=v;
  temp->weight=w;
  temp->next=Graph[u].next;
  Graph[u].next=temp;
}

void display(int n)
{
  for(int i=0; i<n; i++)
  {
    Node* temp=Graph+i;
    while(temp!=NULL)
    {
      cout<<((temp->data)+1)<<"("<<temp->weight<<")"<<"->";
      temp=temp->next;
    }
    cout<<endl;
  }
}

void dijkstra(int n)
{
  int* spset=new int[n];
  for(int i=0; i<n; i++)
  {
    spset[i]=0;
  }
  Heap h(n);
  h.deckey(0, 0);
  while(!h.empty())
  {
    HNode m=h.extractmin();
    spset[m.key]=1;
    cout<<(m.key+1)<<"("<<m.value<<")"<<" ";
    Node* temp=Graph[m.key].next;
    while(temp!=NULL)
    {
      if(spset[temp->data]==0)
      {
        int v=m.value+temp->weight;
        h.deckey(temp->data, v);
      }
      temp=temp->next;
    }
  }
  cout<<endl;
}

int main()
{
  int n, e, u, v, w;
  cout<<"Enter the number of vertices"<<endl;
  cin>>n;
  vlist(n);
  cout<<"Enter the number of edges"<<endl;
  cin>>e;
  cout<<"Enter the edges"<<endl;
  for(int i=0; i<e; i++)
  {
    cin>>u>>v>>w;
    u=u-1;
    v=v-1;
    addedge(u, v, w);
    addedge(v, u, w);
  }
  display(n);
  cout<<"Shortest Path using Dijkstra's Algorithm:"<<endl;
  dijkstra(n);
  return 0;
}

File: Dijkstra-s Algorithm -Python/dijkstra.py
class Node:
    def __init__(self, label=None, weight=0):
        self.label=label;
        self.weight=weight;
        self.next=None;
    def adj(self, v, w):
        temp=Node(v, w);
        temp.next=self.next;
        self.next=temp;
    def printNode(self):
        temp=self;
        while(temp!=None):
            print((temp.label+1), "(", temp.weight, ")", end="  ");
            temp=temp.next;
        print();

def create(n):
    l=[];
    for i in range(0, n):
        l.append(Node(i));
    return l;

def addEdge(l, u, v, w):
    l[u-1].adj((v-1), w);

def display(l):
    for i in l:
        i.printNode();

class Heap:
    def __init__(self, size=0, value=100000):
        self.l=[[i, value] for i in range(0, size)];
        self.pos=[i for i in range(0, size)];
    def empty(self):
        return (len(self.l)==0);
    def parent(self, i):
        if(i!=0):
            return (((i+1)//2)-1);
        return (i);
    def filterUp(self, i):
        p=self.parent(i);
        while(i>0 and self.l[i][1]<self.l[p][1]):
            p=self.parent(i);
            self.pos[self.l[i][0]], self.pos[self.l[p][0]]=self.pos[self.l[p][0]], self.pos[self.l[i][0]];
            self.l[i], self.l[p]=self.l[p], self.l[i];
            i=p;
    def filterDown(self, root):
        min=root;
        left=(2*root)+1;
        right=(2*root)+2;
        if(left<len(self.l) and (self.l[min][1]>self.l[left][1])):
            min=left;
        if(right<len(self.l) and (self.l[min][1]>self.l[right][1])):
            min=right;
        if(min!=root):
            self.pos[self.l[min][0]], self.pos[self.l[root][0]]=self.pos[self.l[root][0]], self.pos[self.l[min][0]];
            self.l[min], self.l[root]=self.l[root], self.l[min];
            self.filterDown(min);
    def decKey(self, key, value=-100000):
        i=self.pos[key];
        if(self.l[i][1]<value):
            return;
        self.l[i][1]=value;
        self.filterUp(i);
    def extractMin(self):
        if(len(self.l)>0):
            min=self.l[0];
            self.l[0]=self.l[-1];
            self.pos[self.l[-1][0]]=0;
            del self.l[-1];
            self.filterDown(0);
        return min;
    def printHeap(self):
        print([i for i in self.l]);

def Dijkstra(graph, source):
    visited=[0 for i in range(0, len(graph))];
    h=Heap(len(graph), 100000);
    h.decKey(source, 0);
    print("Shortest paths using Dijkstra's Algorithm:");
    while(not h.empty()):
        min=h.extractMin();
        print((min[0]+1), "(", min[1], ")");
        visited[min[0]]=1
        temp=graph[min[0]].next;
        while(temp!=None):
            if(not visited[temp.label]):
                w=min[1]+temp.weight;
                h.decKey(temp.label, w);
            temp=temp.next;

if __name__=='__main__':
    n=int(input("Enter the number of vertices:"));
    graph=[];
    graph=create(n);
    e=int(input("Enter the number of edges:"));
    print("Enter the edges:");
    for i in range (0, e):
        s=input();
        edge=[int(x) for x in s.split()];
        addEdge(graph, edge[0], edge[1], edge[2]);
        addEdge(graph, edge[1], edge[0], edge[2]);
    display(graph);
    Dijkstra(graph, 0);
